### PR TITLE
Cluster Features Alternative

### PR DESCRIPTION
**Pull request by Apollon77**
Hey @mfucci,

I played around whole day with ideas around cluster feature and came in the end to that (rough) approach. We add a "supportedFeatures" directly to the Cluster definition. Especially the ClusterExtend (taken over from your PR) only has this supportedFeatures and takes over the features-definition from the base Cluster it extends it by adding the "defined features" supportedFeatures is mandatory in this case.

So options to use:
ClusterBase generates a basic Cluster definition but leaves out the "supportedFeatures". To Add them we can use ClusterExtend. I redesigned the OnOff cluster with this concept. Here also typing is fine. It requires supportedFeatures correctly based on base features.

You can also do further ClusterExtend's and build a chain ... see NetworkCommissioningCluster for that (yes this is then all combinations with 3 feature flags) :-)

The only "issue" that I have left is if someone uses the "Cluster" type directly. When someone does that I did not found a way to have proper typing for "when features is not {} then also supportedFeatures needs to be not {} and vise-versa". Right now someone could provide "features" but without "supportedFeatures".

The alternative would be to forbid that someone uses "Cluster" directly and we always need ClusterBase with an empty CLusterExtend or such, but that feels strange. Many clusters do not have features and so the normal "Cluster" would be enough. Ok we could do a "ClusterWithoutFeature" for these cases ... then "Cluster" would just be the final type to be used but never to be instanciated directly ...

Additionally I have no idea why ESLint states that:

>   168:14  warning  'ClusterExtend' is assigned a value but never used. Allowed unused vars must match /^Matter.*SpecificationV|^_/u  @typescript-eslint/no-unused-vars

... it is in fact used ...

### CODE REVIEW

**Review by JimBuzbee — COMMENTED** [review]
Will this also support Attributes becoming Mandatory based  on the value of another Attribute? i.e not based on a Feature. The ColorControlCluster has an Attribute of "numberOfPrimaries". Based on its value, other Attributes like primary1Intensity, primary2Intensity, primary3Intensity, etc. can become mandatory.

**Comment by Apollon77** [comment]
Hi Jim, no this is not in there explicitly.  It when the numberofPrimaries in general is a fixed attribute (and not dynamically writable) then in fact can also be modeled with it. Let me think about that. 

In fact the bigger open point with the approach here is decoding in case of controllers when I now think about it @mfucci. For controller we would need to have a two pass decoding: decode the cluster with "just mandatory featureMap" (or global attributes in general) and then have a logic per cluster that returns then the relevant feature specific cluster implementation. But on typing side this then becomes dynamic.

**Comment by Apollon77** [comment]
Th more I think about it the more I'm unsure of the approach. When I think generically we need one cluster definitions with ALL attributes, commands and events and such. We need this for generic Attribute knowledge for generic decoding for controller cases. Here accuracy is not that important but the attribute Ute definitions are. 

Then for Cluster server code usage we need to be able to know be correct if we want to be 100% typed, so we need to modify the base "anything in" definition by declaring attributes to mandatory or remove unneeded ones.

With this a "Cluster" is always "anything with too much optional" (or used when no features exists) and ClusterExtend is more a ClusterModify which allows to modify optionality of entries in attributes, commands and events to make them mandatory instead or to remove them.

Then when building cluster servers the dev still uses the correct Cluster instance which was "modified" r the o be correct and we have full type support.

For controller and generic attribute decoding cases we use the Base definition then we can handle anything which is coming in correctly and generic - in this case typing support is difficult in any case but also here after initial attribute decoding we could use the infos we decided (eg festureMap) to instantiate the correct version of it. 

@mfucci WDYT

With that idea I would need that modify typing thing. I have no ideas how to correctly build that.

**Comment by Apollon77** [comment]
@mfucci I adjusted mine PR with your last changes from #87 and also merged master into and such to get it mergeable.

**Review by vves — APPROVED** [review]
Let's do it.

File: packages/matter.js/src/cluster/NetworkCommissioningCluster.ts [hunk 1]
```diff
@@ -224,7 +224,7 @@ const TlvReorderNetworkRequest = TlvObject({
  *
  * @see {@link MatterCoreSpecificationV1_0} § 11.8
  */
-export const NetworkCommissioningCluster = Cluster({
+export const EthernetNetworkCommissioningCluster = Cluster({
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/matter.js/src/cluster/NetworkCommissioningCluster.ts:227]: I would leave it with Metwork to still have the clustername as by specs in it

**Pull request merged**